### PR TITLE
Handle string sets during preview

### PR DESCRIPTION
**Pull request by danielrbradley**
Closes https://github.com/pulumi/pulumi-azure-native/issues/2525

At preview time, we generally just return the inputs back as outputs - as a best-guess as to what might be available after deployment. However, for string sets the input and output SDK shapes are different. Therefore we need to handle this scenario specifically in the "input to output converter". For now, we return a top-level computed object - which is unknown at preview. 

It might be possible to make this known if all elements in the input array are also known but this would be significantly more complex with likely no real-world benefit as the properties within the values still won't be known and they could just extract a variable with their input ids as an array which would be easier to loop over if required rather than extracting the keys from the map.

- Fix preview conversion of StringSet input just gets an unknown output at preview-time rather than returning the input list.
- Add special test case for this type.
- Fix metadata so we have it as the object type in metadata, but an array in the schema.

Metadata diff (when converted to yaml using `cat bin/metadata-compact.json | yj -jy > bin/metadata-compact.yaml`):

```diff
2623,2625c2623
<         type: array
<         items:
<           type: string
---
>         type: object
2657,2659c2655
<         type: array
<         items:
<           type: string
---
>         type: object
7959,7961c7955
<         type: array
<         items:
<           type: string
---
>         type: object
9195,9197c9189
<         type: array
<         items:
<           type: string
---
>         type: object
10645,10647c10637
<         type: array
<         items:
<           type: string
---
>         type: object
12272,12274c12262
<         type: array
<         items:
<           type: string
---
>         type: object
14000,14002c13988
<         type: array
<         items:
<           type: string
---
... snipped ...
```

### CODE REVIEW

File: examples/py-user-assigned-identity/requirements.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+pulumi>=2.0.0,<3.0.0
```
Comment (kpitzen) [inline examples/py-user-assigned-identity/requirements.txt:1]: This seems not correct, right?

Comment (danielrbradley) [reply examples/py-user-assigned-identity/requirements.txt:1]: Hmm, just copied from the other test in there .. guessing it probably just gets ignored/overwritten somewhere in program test.

Comment (mikhailshilkov) [reply examples/py-user-assigned-identity/requirements.txt:1]: Let's change it everywhere

File: examples/py-user-assigned-identity/__main__.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2021, Pulumi Corporation.  All rights reserved.
```
Comment (mikhailshilkov) [inline examples/py-user-assigned-identity/__main__.py:1]: 2023

File: examples/py-user-assigned-identity/Pulumi.yaml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+name: py-user-assigned-identity
+runtime: python
+description: A minimal Azure Python Pulumi program
```
Comment (mikhailshilkov) [inline examples/py-user-assigned-identity/Pulumi.yaml:3]: Please add a description

File: provider/pkg/resources/convert.go [hunk 4]
```diff
@@ -325,33 +325,103 @@ func (k *SdkShapeConverter) PreviewOutputs(inputs resource.PropertyMap,
...
 		} else {
-			result[key] = resource.MakeComputed(resource.NewStringProperty(""))
+			result[key] = k.previewOutputValue(&inputValue, nil)
 		}
 	}
 	return result
 }
 
-func (k *SdkShapeConverter) previewOutputValue(inputValue resource.PropertyValue,
+func (k *SdkShapeConverter) previewOutputValue(inputValue *resource.PropertyValue,
```
Comment (mikhailshilkov) [inline provider/pkg/resources/convert.go:336]: Why this change in `inputValue` type?

File: provider/pkg/resources/convert_test.go [hunk 5]
```diff
@@ -530,10 +530,8 @@ func TestPreviewOutputs(t *testing.T) {
 		"more":  resource.MakeComputed(resource.NewStringProperty("")),
 		"tags":  resource.MakeComputed(resource.NewStringProperty("")),
 		"union": resource.MakeComputed(resource.NewStringProperty("")),
-		"untypedArray": resource.NewArrayProperty([]resource.PropertyValue{
-			resource.NewObjectProperty(resource.PropertyMap{"key1": resource.NewStringProperty("value1")}),
-			resource.NewObjectProperty(resource.PropertyMap{"key1": resource.NewStringProperty("value2")}),
-		}),
+		// If we have no type information for an array or dictionary, we can't do anything with it.
+		"untypedArray": resource.MakeComputed(resource.NewStringProperty("")),
```
Comment (mikhailshilkov) [inline provider/pkg/resources/convert_test.go:534]: Why do we lose properties on arrays but not dicts?

File: provider/pkg/resources/convert.go [hunk 6]
```diff
@@ -325,33 +325,103 @@ func (k *SdkShapeConverter) PreviewOutputs(inputs resource.PropertyMap,
...
 	prop *AzureAPIProperty) resource.PropertyValue {
+	if prop == nil {
```
Comment (mikhailshilkov) [inline provider/pkg/resources/convert.go:338]: It's surprising to see so many changes here without any changes in tests. Why is that? Is this somehow just pure refactoring with no functional changes? Or are we missing tests?

**Review by mikhailshilkov — COMMENTED** [review]
Took the first pass and asked questions. A few high-level remarks:
1. Reviews would be easy if commits were even smaller, with refactorings split into separate ones.
2. The metadata diff in text isn't quite heplful... Can't we commit the file now? Is it still above 100MB?

File: provider/pkg/resources/convert.go (hunk 6 above)
Comment (danielrbradley) [reply provider/pkg/resources/convert.go:338]: There are tests - we're just changing very little of the behaviour. There's one change I made to the tests - untyped arrays will just come back as unknowns during preview as we've got no type information to verify that the output will match the input shape. 

The first iteration of the code had a good few failing tests but I maintained back compatibility where possible.

File: provider/pkg/resources/convert_test.go [hunk 7]
```diff
@@ -542,21 +542,50 @@ func TestPreviewOutputs(t *testing.T) {
...
+		},
+	}
+	inputMap := resource.NewPropertyMapFromMap(inputs)
+	metadata := map[string]AzureAPIProperty{
+		"name": {
+			Type: "string",
+		},
+		"userAssignedIdentities": {
+			Type: "object",
+			AdditionalProperties: &AzureAPIProperty{
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/resources/convert_test.go:557]: Not for the output type, no.

File: provider/pkg/gen/properties.go [hunk 8]
```diff
@@ -246,47 +246,44 @@ func (m *moduleGenerator) genProperty(name string, schema *spec.Schema, context
...
 		IsStringSet:          isStringSet,
 	}
 
 	// Input types only get extra information attached
 	if !isOutput {
 		if m.isEnum(&propertySpec.TypeSpec) {
-			apiProperty = resources.AzureAPIProperty{Type: "string"}
+			apiProperty = resources.AzureAPIProperty{Type: "string", ForceNew: forceNew}
 		} else {
-			// Set additional properties when it's an input
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/gen/properties.go:276]: Ah, before squashing there was a step where I took the assignments out of this conditional block - must have got lost when putting them back.

File: provider/pkg/resources/convert.go (hunk 6 above)
Comment (mikhailshilkov) [reply provider/pkg/resources/convert.go:338]: Thank you, that's great. Can you elaborate on why you needed to make these changes then?

File: provider/pkg/resources/convert_test.go (hunk 5 above)
Comment (mikhailshilkov) [reply provider/pkg/resources/convert_test.go:534]: I'm a bit scared that something relied on the past behavior, given there was a test for it, and that we may break it inadvertently. Should we dig up a PR which introduced the original test and see if anything may be affected?

File: provider/pkg/resources/convert.go [hunk 9]
```diff
@@ -325,33 +325,103 @@ func (k *SdkShapeConverter) PreviewOutputs(inputs resource.PropertyMap,
...
+		return resource.MakeComputed(resource.NewStringProperty(""))
+	}
+	if prop.Const != nil {
+		return resource.NewStringProperty(prop.Const.(string))
+	}
+	if inputValue == nil {
```
Comment (mikhailshilkov) [inline provider/pkg/resources/convert.go:344]: What does this case represent? It looks like this will never be true in the current implementation? (you never pass a nil)

File: provider/pkg/resources/convert.go (hunk 6 above)
Comment (danielrbradley) [reply provider/pkg/resources/convert.go:338]: The changes are required for the fall-through case - we were defaulting to returning the input if none of the conditions matched. Now we're being more stringent on our type checking and then falling back to returning a computed property if we can't guarentee the input valyue matches the output shape.

File: provider/pkg/resources/convert.go (hunk 9 above)
Comment (danielrbradley) [reply provider/pkg/resources/convert.go:344]: Yeah, pretty sure we don't ever get to a nil here, but the type is a pointer, so seems safest to include the check to avoid any chance of runtime panics - especially if this is gets called differently in the future.

File: provider/pkg/resources/convert.go (hunk 4 above)
Comment (danielrbradley) [reply provider/pkg/resources/convert.go]: Now refactored to put it back to a non-pointer in a follow-up commit.

File: provider/pkg/gen/properties.go [hunk 10]
```diff
@@ -246,47 +246,44 @@ func (m *moduleGenerator) genProperty(name string, schema *spec.Schema, context
...
+		WillReplaceOnChanges: forceNew,
+	}
+
 	if isStringSet {
-		typeSpec.Type = "array"
-		typeSpec.AdditionalProperties = nil
-		typeSpec.Items = &pschema.TypeSpec{
+		propertySpec.Type = "array"
+		propertySpec.AdditionalProperties = nil
+		propertySpec.Items = &pschema.TypeSpec{
```
[the remark this answers is not in the document]
Comment (danielrbradley) [reply provider/pkg/gen/properties.go]: I've now renamed these variables which makes the whole section much easier to understand.

**Comment by danielrbradley** [comment]
@mikhailshilkov I think I've now addressed all your feedback inline.

1. With this work this wasn't much of a refactoring job - more of a straight up rewrite so I'm afraid there wasn't really any intermediate steps when reworking the core logic.
2. We _could_ start committing the metadata again, though I think it would be of limited value as-is. We'd probably want to commit it without compaction - which would push the size up, and the diffs would likely often be hidden due to being in a single file. Perhaps we could explore this as a separate topic at some other point - perhaps splitting into smaller files (per service maybe) or even using a different format to make the diffs better.

**Comment by danielrbradley** [comment]
Just for completeness, here's what we see when we run the new tests against the old preview implementation.

- All of the valid cases (except string set) function the same.
- All of the mismatch cases now function differently (these weren't really tested before).
- We're now a little smater with const types - and will just return the const value at preview time.

image

**Review by kpitzen — APPROVED** [review]
